synapse/http/matrixfederationclient.py

```diff
@@ -902,7 +935,7 @@ async def put_json(
             _sec_timeout = self.default_timeout
 
         if parser is None:
-            parser = JsonParser()
+            parser = cast(ByteParser[T], JsonParser())
```

Comment:
mypy didn't like not having this cast:

```
synapse/http/matrixfederationclient.py:938: error: Incompatible types in assignment (expression has type "JsonParser", variable has type "Optional[ByteParser[T]]")  [assignment]
synapse/http/matrixfederationclient.py:946: error: Argument "parser" to "_handle_response" has incompatible type "Optional[ByteParser[T]]"; expected "ByteParser[T]"  [arg-type]
```

Reply:
I would guess this is because JsonParser is a `ByteParser[Union[JsonDict, list]]`and not a `ByteParser[T]`. Mypy might be happier if we make two separate calls to `handle_response`. But I don't think we need to bother.